File: .agents/skills/explore-codebase-ast/scripts/explore.js
const fs = require('fs');
const path = require('path');

const rootPath = process.argv[2];
const extensions = process.argv[3].split(',').map(ext => ext.trim().startsWith('.') ? ext.trim() : `.${ext.trim()}`);
const maxDepth = parseInt(process.argv[4], 10);

console.log(`## Architecture Map for \`${rootPath}\``);
console.log(`> [!WARNING]\n> Using fallback Node.js script (Regex-based). For true polyglot AST parsing, use Python with \`tree-sitter-languages\`.\n`);

function extractEntities(content, ext) {
    const entities = [];
    
    if (['.js', '.ts', '.java', '.cs', '.php'].includes(ext)) {
        // Naive Class matching
        const classRegex = /class\s+([A-Za-z0-9_]+)/g;
        let match;
        while ((match = classRegex.exec(content)) !== null) {
            entities.push(`- Class: ${match[1]}`);
        }
        
        // Naive Interface matching
        const interfaceRegex = /interface\s+([A-Za-z0-9_]+)/g;
        while ((match = interfaceRegex.exec(content)) !== null) {
            entities.push(`- Interface: ${match[1]}`);
        }
    }
    
    return entities;
}

function walkDir(currentPath, currentDepth) {
    if (currentDepth > maxDepth) return;

    let items;
    try {
        items = fs.readdirSync(currentPath);
    } catch (e) {
        return;
    }

    // Sort to keep output consistent
    items.sort();

    for (const item of items) {
        if (['.git', 'node_modules', 'venv', '__pycache__', 'dist', 'build', 'out'].includes(item)) {
            continue;
        }

        const itemPath = path.join(currentPath, item);
        const indent = '  '.repeat(currentDepth);

        let stat;
        try {
            stat = fs.statSync(itemPath);
        } catch (e) {
            continue;
        }

        if (stat.isDirectory()) {
            console.log(`${indent}📁 ${item}/`);
            walkDir(itemPath, currentDepth + 1);
        } else {
            const ext = path.extname(item);
            if (extensions.includes(ext)) {
                console.log(`${indent}📄 ${item}`);
                const content = fs.readFileSync(itemPath, 'utf8');
                const entities = extractEntities(content, ext);
                for (const entity of entities) {
                    console.log(`${indent}  ${entity}`);
                }
            }
        }
    }
}

walkDir(rootPath, 0);

File: .agents/skills/explore-codebase-ast/scripts/explore.py
import os
import sys
import argparse

try:
    from tree_sitter_languages import get_parser
    HAS_TREE_SITTER = True
except ImportError:
    HAS_TREE_SITTER = False

# Mapping common extensions to tree-sitter language names
EXT_TO_LANG = {
    '.py': 'python',
    '.js': 'javascript',
    '.ts': 'typescript',
    '.java': 'java',
    '.cs': 'c_sharp',
    '.cpp': 'cpp',
    '.c': 'c',
    '.go': 'go',
    '.rb': 'ruby',
    '.rs': 'rust',
    '.php': 'php',
}

def parse_file(filepath, ext):
    lang_name = EXT_TO_LANG.get(ext)
    if not lang_name or not HAS_TREE_SITTER:
        return []

    try:
        parser = get_parser(lang_name)
        with open(filepath, 'r', encoding='utf-8') as f:
            code = f.read()
        tree = parser.parse(bytes(code, "utf8"))
        
        entities = []
        
        def traverse(node):
            if node.type in ['class_declaration', 'class_definition']:
                name_node = next((c for c in node.children if c.type == 'identifier'), None)
                if name_node:
                    name = code[name_node.start_byte:name_node.end_byte]
                    entities.append(f"- Class: {name}")
            elif node.type in ['function_declaration', 'function_definition', 'method_declaration', 'method_definition']:
                name_node = next((c for c in node.children if c.type in ['identifier', 'property_identifier']), None)
                if name_node:
                    name = code[name_node.start_byte:name_node.end_byte]
                    entities.append(f"  - Method: {name}")
            elif node.type in ['interface_declaration']:
                name_node = next((c for c in node.children if c.type == 'type_identifier'), None)
                if name_node:
                    name = code[name_node.start_byte:name_node.end_byte]
                    entities.append(f"- Interface: {name}")

            for child in node.children:
                traverse(child)

        traverse(tree.root_node)
        return entities
    except Exception as e:
        return [f"  - [Error parsing AST: {str(e)}]"]

def explore(root_path, target_extensions, max_depth):
    target_exts = [e.strip() if e.startswith('.') else f".{e.strip()}" for e in target_extensions.split(',')]
    
    print(f"## AST Architecture Map for `{root_path}`")
    if not HAS_TREE_SITTER:
        print("> [!WARNING]\n> `tree_sitter_languages` not installed. Showing only files without AST.\n> Run: `pip install tree-sitter tree-sitter-languages`\n")
    
    def walk_dir(current_path, current_depth):
        if current_depth > max_depth:
            return

        try:
            items = sorted(os.listdir(current_path))
        except PermissionError:
            return

        for item in items:
            if item in ['.git', 'node_modules', 'venv', '__pycache__', 'dist', 'build', 'out']:
                continue

            item_path = os.path.join(current_path, item)
            indent = "  " * current_depth
            
            if os.path.isdir(item_path):
                print(f"{indent}📁 {item}/")
                walk_dir(item_path, current_depth + 1)
            else:
                _, ext = os.path.splitext(item)
                if ext in target_exts:
                    print(f"{indent}📄 {item}")
                    if HAS_TREE_SITTER:
                        entities = parse_file(item_path, ext)
                        for entity in entities:
                            print(f"{indent}  {entity}")

    walk_dir(root_path, 0)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Explore AST of a codebase.")
    parser.add_argument("path", help="Root directory to explore")
    parser.add_argument("extensions", help="Comma-separated list of extensions (e.g. .ts,.js)")
    parser.add_argument("depth", type=int, help="Maximum depth to traverse")
    args = parser.parse_args()
    
    explore(args.path, args.extensions, args.depth)
